Make log() recognize some exact values

gschintgen opened the pull request:
With these changes, `log(a+b*I)` will automatically evaluate if its argument has a complex argument that is a rational multiple of pi (with a denominator in {3, 4, 5, 6, 8, 10, 12}.)

#### References to other Issues or PRs
Fixes #4612 

#### Brief description of what is fixed or changed
A constants table is added to log.eval() in order to detect some specific log arguments. This is similar to the approach used by the inverse trigonometric functions.
~~Only arguments leading to a purely imaginary result are evaluated automatically.~~

#### Other comments
~~I refrained from adding support for other angles (pi/5, pi/8, pi/10, pi/12) primarily because some form of pre-processing would be necessary: I could add `sqrt(2)/4 + sqrt(6)/4 - I*sqrt(2)/4 + I*sqrt(6)/4` to the table, but then `log(cos(pi/12) + I*sin(pi/12))` still wouldn't be recognized since it evaluates to `log(sqrt(2)/4 + sqrt(6)/4 + I*(-sqrt(2)/4 + sqrt(6)/4))` (with `I` factored out).
Obviously, potentially slow operation (simplify etc.) are to be avoided, since most of the time the log argument won't be listed in the table.
Is there some kind of pre-processing that would be acceptable? Maybe just `re()` and `im()` or `expand_mul(deep=False)` for each of the top-level terms of the Add?~~
Any review or advice is most welcome.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* functions
  * log() now recognizes algebraic values having a complex argument of `I*n*pi/3`, `I*n*pi/4`, `I*n*pi/5`,  `I*n*pi/6`, `I*n*pi/8`, `I*n*pi/10` or `I*n*pi/12`
<!-- END RELEASE NOTES -->

gschintgen committed:
fix test for issue 6056 (nicer result even without simplify)

gschintgen committed:
Make log() recognize some exact values

With these changes, log(a+b*I) will automatically evaluate
if the result is of the form I*pi/d, where d is in {3, 4, 6}.

jksuom commented on sympy/functions/elementary/tests/test_exponential.py:
```diff
@@ -191,6 +191,13 @@ def test_log_values():
+    assert log(sqrt(3)/2 + I*S.Half) == I*pi/6
+    assert log(S.Half - I*sqrt(3)/2) == -I*pi/3
+    assert log(sqrt(2)/2 + I*sqrt(2)/2) == I*pi/4
+    assert log(-S.Half + I*sqrt(3)/2) == I*2*pi/3
```
Would it be possible to extend the algorithm so that, for example, `log(-1 + I*sqrt(3))` would give `log(2) + I*2*pi/3`?

gschintgen replied:
Yes, it'd definitely be possible to do slightly more advanced simplifications automatically. I'm mainly wondering about the following though and I'd be very happy to read your thoughts on it:

Is such an automatic transformation always wanted? What would be the limits? Hypothetically: `log(-8 + 8*I) = log(2*sqrt(2)) - 3*I*pi/4` or even `log(2*sqrt(6)*I-sqrt(8)) = log(4*sqrt(2)) + 2*I*pi/3`. The presentation of the expressions would change significantly but there'd still be a logarithm involved. (Currently these transformations are left to `expand_complex`.)

Possibly the automatic evaluation could be limited to cases where the log argument is an integer? (Using `as_content_primitive` for factoring out the constant?)

In any case I think I'll also add the necessary processing for recognizing other special angles.

jksuom replied:
It might suffice to consider those arguments that `pure_complex` will resolve, at least to begin with. Then there could be a lookup table for some values of `y/x` that would give the imaginary part of the logarithm.

gschintgen replied:
Thanks for your reply. As far as I can tell, `pure_complex` by itself won't be of much help: it bails as soon as a radical is involved.
```
pure_complex(sqrt(2)/2+I*sqrt(2)/2)
pure_complex(sqrt(3)+I)
```
So it'll have to be `re()` and `im()`.
I'll do some more experiments locally before pushing any changes. (I'm marking this PR "WIP" for now.)

smichr replied:
> (2*sqrt(6)*I-sqrt(8))

FYI: `factor_terms` can be used to get a canonical surd expression:
```python
>>> factor_terms(2*sqrt(6)*I-sqrt(8))
2*(-sqrt(2) + sqrt(6)*I)
>>> factor_terms(2*sqrt(6)*I-sqrt(8), radical=True)
2*sqrt(2)*(-1 + sqrt(3)*I)
```

gschintgen replied:
Thanks! Using `factor_terms` as a canonicalization tool for surd expressions is very interesting. I did some quick experiments/research and noticed that `cancel` already calls `factor_terms` as a first step anyway, **but** my current code tries to specifically match `a+b*I` which then fails for the form given by `factor_terms` as output due to the additional multiplicative factor(s). As a result the current PR code evaluates the logarithm just fine for `2*sqrt(6)*I-sqrt(8)` but fails (by design) if the same expression is given in a factored form.

I'll see if I can enlarge the set of properly recognized (and equivalent) expressions without incurring too much of a computational penalty.

smichr replied:
So I suspect you solved this by removing the multiplicative coefficient here `coeff, arg_ = arg.as_independent(I, as_Add=False)`? i.e. does it solve `2*sqrt(6)*I-sqrt(8)` in factored form?

gschintgen replied:
Here's the current output:
```
In [7]: log(2*sqrt(6)*I-sqrt(8))
Out[7]:
            2⋅ⅈ⋅π
log(4⋅√2) + ─────
              3

In [8]: factor_terms(2*sqrt(6)*I-sqrt(8))
Out[8]: 2⋅(-√2 + √6⋅ⅈ)

In [9]: log(_)
Out[9]:
            2⋅ⅈ⋅π
log(4⋅√2) + ─────
              3

In [10]: factor_terms(2*sqrt(6)*I-sqrt(8), radical=True)
Out[10]: 2⋅√2⋅(-1 + √3⋅ⅈ)

In [11]: log(_)
Out[11]:
            2⋅ⅈ⋅π
log(4⋅√2) + ─────
              3

```
So yes, the first `.as_independent` extracts an already present factor and reintegrates it later on in the modulus. I also tried adding a call to `factor_terms` as first step so as to have a simpler `r_ + i_*I` for `cancel(i_/r_)` to take care of, but as far as I could tell there was no discernible benefit (using `timeit`). I also tried to canonicalize using only `factor_terms` and `signsimp` (i.e. without `cancel`), but that didn't work out.

gschintgen committed:
Adapt a solve() test to the improved log() values

Both asserts in test_issue_2840_8155() were failing. This commit
updates the expected results to the simplified log() results.
(The numeric values of the solutions did not change, only their
presentation.)

gschintgen committed:
Improve exact value recognition for log(); add tests

gschintgen committed:
Add an exact value example to log() docstring

oscargus commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -1843,19 +1837,16 @@ def test_issue_8828():
-        0, -pi, pi, 2*I*(log(2) - log(-sqrt(3) - I)), 2*I*(log(2) -
-        log(-sqrt(3) + I)), 2*I*(log(2) - log(sqrt(3) - I)), 2*I*(log(2) -
-        log(sqrt(3) + I))]
+        0, -5*pi/3, -pi, -pi/3, pi/3, pi, 5*pi/3]
```
This is a very good example why this PR is definitely worthwhile!

gschintgen committed:
Recognize exact log values only if arg.is_algebraic

re(arg) and im(arg) on some complicated expressions can lead to
uncontrolled/runaway computations. Case in point: without this
restriction, the meijerg=True integrations in test_messy() will not
finish in a reasonable time.

jksuom commented on sympy/functions/elementary/exponential.py:
```diff
@@ -576,22 +581,37 @@ def eval(cls, arg, base=None):
+            # multiples of pi
+            r_ = re(arg)
+            i_ = im(arg)
+            t = (i_/r_).cancel()
```
The case `r_ == 0` should be handled somewhere.

gschintgen replied:
Thanks for reviewing!

Indeed, explicitly handling `r_==0` would be cleaner (I'll modify the code.). Currently `i_/r_ = zoo` for `arg=5*I` and `zoo` isn't in the table, so it falls through to the next code block, which handles it.

It may be best to also change the order of the code blocks, i.e. put the exact values code at the end of `eval()`. That way most of the purely imaginary arguments will already have been handled (at least rational or infinite multiples of `I`).  And that leaves us with the question of how much *automatic* simplification/evaluation is actually wanted. I'm not sure the following would always be appreciated: `log(I*sqrt(2)) = log(sqrt(2)) + I*pi/2`. (Result arguably more complicated than the input.) But then again it's already implemented for rational coefficients.

With the current PR code, [one of the test cases](https://github.com/sympy/sympy/blob/a2d65f2a9c6c774404d602a941c09008e100025a/sympy/series/tests/test_formal.py#L276) is failing rather spectacularly (overly complex result) and the only relevant `log` calls seem to be `log(1+I)` and `log(1-I)`.
I didn't notice before pushing because it's one of the slow tests that are skipped by `test()`.
I'll try to track it down.

gschintgen replied:
I think both issues should now be resolved with the latest commits.

jksuom commented:
The error is a consequence of the automatic rewrite of some logarithms. (There could have been more of them.)

The formal power series generator finds the coefficients by using `rsolve` to solve a recursion equation. `rsolve` transforms the equation into a set of algebraic relations.
https://github.com/sympy/sympy/blob/a2d65f2a9c6c774404d602a941c09008e100025a/sympy/solvers/recurr.py#L816-L820

It is `limit` that will give an unexpected result here, confusing the algebraic solver:
```
>>> e = (1 - I)**n
>>> e.limit(n, 1)
-(-1)**(3/4)*sqrt(2)
>>> e.subs(n, 1)
1 - I
```
The reason is that, at some point, `limit` writes `e` in the form `exp(n*log(1 - I))` where
```
>>> log(1 - I)
log(sqrt(2)) - I*pi/4
```
and  then `exp(-I*pi/4)` becomes `-(-1)**(3/4)`.

It might be possible to fix issues like this by adding to `exp`a similar evaluation table transforming `exp(-I*pi/4)` back to `1 - I`

gschintgen commented:
> It is `limit` that will give an unexpected result here, confusing the algebraic solver:
> 
> ```
> >>> e = (1 - I)**n
> >>> e.limit(n, 1)
> -(-1)**(3/4)*sqrt(2)
> >>> e.subs(n, 1)
> 1 - I
> ```
> 
> The reason is that, at some point, `limit` writes `e` in the form `exp(n*log(1 - I))` where
> 
> ```
> >>> log(1 - I)
> log(sqrt(2)) - I*pi/4
> ```
> 
> and then `exp(-I*pi/4)` becomes `-(-1)**(3/4)`.
> 
> It might be possible to fix issues like this by adding to `exp`a similar evaluation table transforming `exp(-I*pi/4)` back to `(1 - I)/sqrt(2)`.

Thanks for the detailed analysis!

The more I think about this, the more I'm wondering if some of the proposed auto-evals won't turn out to be harmful (as outlined in the enlightening [wiki article on this subject](https://github.com/sympy/sympy/wiki/Automatic-Simplification)). E.g. high-school computations like the following currently work just fine and as expected:
```
>>> exp(I*pi/4)/exp(I*pi/12)
exp(I*pi/6)
```
If a result in algebraic form is wanted, `expand_complex` does it just fine. But as soon as the exponentials are auto-evaluated the user will have to actually work against the simplifier. The same also holds for other contexts. (E.g. `sin(pi/40)/cos(pi/40)*cot(pi/40)` is not very helpful right now... A purely formal/symbolic manipulation is somewhat sabotaged. Though in such a case it'd be best practice to use variables, then `subs`titute.)
Wouldn't it be better to restrict the *automatic* log evaluation to expressions whose the result is *definitely* simpler, i.e. a single imaginary term? (I'd gladly add a note on `expand_complex` to `log`'s docstring in this case, including an example! It's not obvious for new users.)

I've just noticed that wolframalpha is also somewhat more conservative: It applies the automatic log evaluation in the case `log(sqrt(3)/2+I/2)` (giving an "exact result"), but doesn't apply it to `log(sqrt(3)+I)` (no "exact result" is given, but the expanded form is listed as "alternate form").

gschintgen commented:
Quick update: With the following crude patch to auto-evaluate `exp` I'm getting even more test failures... I didn't inspect them all, a few seemed rather benign. (slow tests not tested)
```
--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -263,6 +263,13 @@ def eval(cls, arg):
                             return -S.ImaginaryUnit
                         elif ask(Q.odd(coeff + S.Half)):
                             return S.ImaginaryUnit
+                    elif coeff.is_Rational:
+                        from sympy.functions.elementary.trigonometric \
+                                import cos, sin
+                        if coeff.q in [3, 4, 6]:
+                            I = S.ImaginaryUnit
+                            return cos(coeff*S.Pi) + I*sin(coeff*S.Pi)
```

```
______________________________________________ sympy/core/tests/test_arit.py:test_mul_coeff _______________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/core/tests/test_arit.py", line 1896, in test_mul_coeff
    assert p**2*x*p*y*p*x*p**2 == x**2*y
AssertionError
___________________________________________________________________________________________________________________________________________
____________________________ sympy/physics/quantum/tests/test_identitysearch.py:test_bfs_identity_search_xfail ____________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_identitysearch.py", line 497, in test_bfs_identity_search_xfail
    assert bfs_identity_search(gate_list, 1, max_depth=3) == id_set
AssertionError
___________________________________________________________________________________________________________________________________________
___________________________________ sympy/physics/quantum/tests/test_spin.py:test_represent_spin_states ___________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_spin.py", line 70, in test_represent_spin_states
    JzKet(S(1)/2, S(1)/2), basis=Jy) == sqrt(2)*Matrix([-1, I])/2
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________ sympy/physics/quantum/tests/test_spin.py:test_represent_uncoupled_states _________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_spin.py", line 110, in test_represent_uncoupled_states
    Matrix([-I, 0, 0, 0])
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________________ sympy/physics/quantum/tests/test_spin.py:test_rotation_d _________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/physics/quantum/tests/test_spin.py", line 3595, in test_rotation_d
    S(1)/2, S(1)/2, S(1)/2, pi/2, pi/2, pi/2).doit() == -I*sqrt(2)/2
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________________ sympy/simplify/tests/test_hyperexpand.py:test_branch_bug _________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_hyperexpand.py", line 25, in test_branch_bug
    + sqrt(pi)*erf(sqrt(z))/(5*sqrt(z))
AssertionError
___________________________________________________________________________________________________________________________________________
____________________________________________ sympy/simplify/tests/test_powsimp.py:test_powsimp ____________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_powsimp.py", line 99, in test_powsimp
    assert str( powsimp(exp(I*pi/3)*root(-1,3)) ) == '(-1)**(2/3)'
AssertionError
___________________________________________________________________________________________________________________________________________
________________________________________ sympy/simplify/tests/test_simplify.py:test_simplify_other ________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_simplify.py", line 155, in test_simplify_other
    assert simplify(ans) == -(-1)**(S(3)/4)*sqrt(pi)/sqrt(t)
AssertionError
___________________________________________________________________________________________________________________________________________
__________________________________________ sympy/simplify/tests/test_simplify.py:test_besselsimp __________________________________________
Traceback (most recent call last):
  File "/root/sympy/sympy/simplify/tests/test_simplify.py", line 603, in test_besselsimp
    besselj(a, sqrt(x)) * cos(sqrt(x))
AssertionError
```

jksuom commented:
> With the following crude patch to auto-evaluate exp I'm getting even more test failures.

I guess that is unavoidable. Many of the failures can probably be fixed by editing the test without changing its mathematical content. But if that is not possible, then automatic evaluation should be avoided. Perhaps some explicit rewrite method could be defined instead (`_eval_rewrite_as_...`).

gschintgen committed:
Refactor log.eval for exact values (rational multiples of pi)

- Only basic matching of a+b*I is done, in order to avoid potentially
expensive expansion by re, im or as_real_imag.
- log(b*I) with algebraic b is now evaluated/expanded automatically.
- For some angles two different tangent expressions (im/re) are
recognized: the completely simplified value and the one obtained by
(sin/cos).cancel().
- In exp._eval_nseries ratsimp may introduce unexpanded (-1)**Rational
which ultimately can lead to unsimplified values for limit() or fps().
These powers are now expanded in some cases (exponent n/3, n/4, n/6).

gschintgen committed:
log(): Add more tests for auto-eval of exact values

gschintgen commented:
> 
> 
> > With the following crude patch to auto-evaluate exp I'm getting even more test failures.
> 
> I guess that is unavoidable. Many of the failures can probably be fixed by editing the test without changing its mathematical content. But if that is not possible, then automatic evaluation should be avoided. Perhaps some explicit rewrite method could be defined instead (`_eval_rewrite_as_...`).

While most test failures for auto-eval of exp() were "benign" in the sense, that they didn't lead to complete failure or obviously wrong results, they /did/ exhibit more complicated output as tested for. So it would have been necessary to edit the relevant code in various modules, not the tests.

As for the `_eval_rewrite_as_...`, there is already `exp._eval_rewrite_as_sqrt`. I tried applying it selectively (only for pi/3, pi/4, pi/6) in `exp._eval_nseries`, but this led to other failures. Another approach proved more fruitful: in `exp._eval_nseries` I could remove the problematic `(-1)**Rational` by selectively expanding it. In my local testing that eliminated the test failure.

I'd be grateful if you could review the new changes. (Some more explanations can be found in the commit message.)

smichr committed:
Merge branch 'master' into fix4612-log-values-v1

smichr committed:
Update test_exponential.py

gschintgen committed:
log() exact values: handle multiplicative coefficient

- Properly handle arguments given as coeff*(r + i*I)
  instead of just r + i*I.
- Add some more tests in test_log_exact().
- Fix an unfortunate typo in log()'s atan_table.

oscargus commented on sympy/functions/elementary/exponential.py:
```diff
@@ -578,8 +590,9 @@ def eval(cls, arg, base=None):
             return arg._eval_func(cls)
 
         if arg.is_number:
+            I = S.ImaginaryUnit
```
If you make some more changes (seems like a waste to run the tests again etc for this small comment), but it may make sense to define this just before the `if arg.is_number` and then use it even in the lines before it is redefined (and not redefine it).

smichr replied:
What do you think, @gschintgen ? Are there more changes? Everything except for this looks fine to me. This detail could be handled in some other PR with other changes.

gschintgen commented:
I don't plan any more changes for this PR, but I do plan other refinements in exponential.py, so I could integrate it then.

On 28 July 2019 16:30:12 CEST, Christopher Smith <notifications@github.com> wrote:
>smichr commented on this pull request.
>
>
>
>> @@ -578,8 +590,9 @@ def eval(cls, arg, base=None):
>             return arg._eval_func(cls)
> 
>         if arg.is_number:
>+            I = S.ImaginaryUnit
>
>What do you think, @gschintgen ? Are there more changes? Everything
>except for this looks fine to me. This detail could be handled in some
>other PR with other changes.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/sympy/sympy/pull/17201#discussion_r308004393

oscargus commented:
Let's go! Thanks @gschintgen!
